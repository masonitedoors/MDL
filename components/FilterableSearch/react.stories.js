// import React from 'react'
// import { storiesOf } from '@storybook/react'
// import { withKnobs } from '@storybook/addon-knobs'
// import { withState } from '@dump247/storybook-state'
// import FilterableSearch from './react'
// import Readme from './react.README.md'

// storiesOf('FilterableSearch', module)
//   .addParameters({ readme: { sidebar: Readme } })
//   .addDecorator(withKnobs)
//   .add(
//     'default',
//     withState({
//       value: '',
//       filterChoices: [
//         {
//           label: 'Filter 1',
//           value: 'FILTER1',
//           checked: true,
//         },
//         {
//           label: 'Filter 2',
//           value: 'FILTER2',
//           checked: false,
//         },
//       ],
//     })(({ store }) => (
//       <FilterableSearch
//         onFilterChange={checkedChoice => {
//           store.set({
//             ...store.state,
//             filterChoices: store.state.filterChoices.map(choice =>
//               (choice.value === checkedChoice.value
//                 ? {
//                   ...choice,
//                   checked: !choice.checked,
//                 }
//                 : choice)),
//           })
//         }}
//         value={store.state.value}
//         onInputChange={value => {
//           console.log('handleInputChange invoked, returns: ', value)
//           store.set({
//             value,
//           })
//         }}
//         onSubmit={value => {
//           console.log('handleSearch callback invoked, returns:', value)
//         }}
//         filterChoices={store.state.filterChoices}
//       />
//     )),
//   )
//   .add(
//     'without onSubmit',
//     withState({
//       value: '',
//       filterChoices: [
//         {
//           label: 'Filter 1',
//           value: 'FILTER1',
//           checked: true,
//         },
//         {
//           label: 'Filter 2',
//           value: 'FILTER2',
//           checked: false,
//         },
//       ],
//     })(({ store }) => (
//       <FilterableSearch
//         placeholder="Filter"
//         onFilterChange={checkedChoice => {
//           store.set({
//             ...store.state,
//             filterChoices: store.state.filterChoices.map(choice =>
//               (choice.value === checkedChoice.value
//                 ? { ...choice, checked: !choice.checked }
//                 : choice)),
//           })
//         }}
//         value={store.state.value}
//         onInputChange={value => {
//           console.log('handleInputChange invoked, returns: ', value)
//           store.set({
//             value,
//           })
//         }}
//         filterChoices={store.state.filterChoices}
//       />
//     )),
//   )
