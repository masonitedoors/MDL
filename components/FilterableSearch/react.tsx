import React, { PureComponent } from 'react'
import classNames from 'classnames/bind'
import { mFilter, mChevronDown } from '@masonite/svg-icons'
import { React as TextField } from 'components/TextField'
import { React as Checkbox } from 'components/Checkbox'
import style from './style.module.scss'

const cx = classNames.bind(style)

interface FilterableSearchState {
  showFilters: boolean
}

export interface FilterableSearchProps {
  buttonLabel?: string
  dropdownIcon?: 'chevronDown' | 'filter'
  onFilterChange: any
  onInputChange?: any
  onSubmit?: any
  value: string
  filterChoices: {
    label: string
    value: string
    checked: boolean
  }[]
  placeholder: string
}

class FilterableSearch extends PureComponent<FilterableSearchProps> {
  state: FilterableSearchState = {
    showFilters: false,
  }

  dropdownIcons: Record<string, string> = {
    chevronDown: mChevronDown,
    filter: mFilter,
  }

  dropdownMenuRef = React.createRef<HTMLDivElement>()

  componentDidMount() {
    document.body.addEventListener('click', this.handleClick)
  }

  componentWillUnmount() {
    document.body.removeEventListener('click', this.handleClick)
  }

  handleClick = ({ target }) => {
    const {
      dropdownMenuRef,
      state: { showFilters },
    } = this

    dropdownMenuRef.current
      && showFilters
      && !dropdownMenuRef.current.contains(target)
      && this.setState({ showFilters: false })
  }

  render() {
    const {
      buttonLabel = 'Search',
      dropdownIcon = 'chevronDown',
      onFilterChange,
      onInputChange,
      onSubmit,
      filterChoices = [],
      placeholder = 'Search',
      value,
    } = this.props

    const { showFilters } = this.state

    const selectedDropdownIcon = dropdownIcon && this.dropdownIcons[dropdownIcon]

    const SearchButton = () => (
      <button
        type="button"
        onClick={() => onSubmit && onSubmit(value)}
        className={cx(['filterable-search__btn'])}
      >
        {buttonLabel}
      </button>
    )

    const DropdownToggle = () => (
      <button
        type="button"
        className={cx([
          'filterable-search__dropdown-toggle',
          { 'filterable-search__dropdown-toggle--active': showFilters },
        ])}
        onClick={() =>
          this.setState((prevState: FilterableSearchState) => ({
            showFilters: !prevState.showFilters,
          }))
        }
      >
        <div
          className={cx([
            'filterable-search__dropdown-toggle-icon',
            { 'filterable-search__dropdown-toggle-icon--active': showFilters },
          ])}
          dangerouslySetInnerHTML={{ __html: selectedDropdownIcon }}
        />
      </button>
    )

    const DropdownMenu = ({ showFilters }: any) => (
      <ul
        className={cx([
          'filterable-search__dropdown-menu',
          { 'filterable-search__dropdown-menu--open': showFilters },
        ])}
      >
        {filterChoices.map(({ label, value, checked }, index) => (
          <li className={cx(['filterable-search__dropdown-menu-item'])} key={value}>
            {/* eslint-disable-next-line */}
            <label
              className={cx(['filterable-search__dropdown-menu-item-label'])}
              htmlFor={`${value}${index}`}
            >
              <Checkbox
                checked={checked}
                onChange={() => onFilterChange({ label, value, checked })}
              />
              <span className={cx(['filterable-search__dropdown-menu-item-label-text'])}>
                {label}
              </span>
            </label>
          </li>
        ))}
      </ul>
    )

    return (
      <div className={cx(['filterable-search'])}>
        <TextField
          placeholder={placeholder}
          onChange={(value: string) => onInputChange(value)}
          onKeyDown={(ev: KeyboardEvent) => {
            if (ev.key === 'Enter' && onSubmit) onSubmit((ev.target as HTMLInputElement).value)
          }}
          value={value}
        />
        {onSubmit && <SearchButton />}
        <div ref={this.dropdownMenuRef}>
          <DropdownToggle />
          <DropdownMenu showFilters={showFilters} />
        </div>
      </div>
    )
  }
}

export default FilterableSearch
