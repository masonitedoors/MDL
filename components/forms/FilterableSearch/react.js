import React, { PureComponent } from 'react'
import classNames from 'classnames/bind'
import { mFilter, mChevronDown } from '@masonite/svg-icons'
import { React as TextField } from 'components/forms/TextField'
import { React as Checkbox } from 'components/forms/Checkbox'
import PropTypes from 'prop-types'
import style from './style.module.scss'

const cx = classNames.bind(style)

class FilterableSearch extends PureComponent {
  state = {
    showFilters: false,
  }

  dropdownIcons = {
    chevronDown: mChevronDown,
    filter: mFilter,
  }

  dropdownMenuRef = React.createRef()

  componentDidMount() {
    document.body.addEventListener('click', this.handleClick)
  }

  componentWillUnmount() {
    document.body.removeEventListener('click', this.handleClick)
  }

  handleClick = ({ target }) => {
    const { showFilters } = this.state
    const dropdownMenu = this.dropdownMenuRef.current

    if (showFilters && !dropdownMenu.contains(target)) {
      this.setState({ showFilters: false })
    }
  }

  render() {
    const {
      buttonLabel,
      dropdownIcon,
      onFilterChange,
      onInputChange,
      onSubmit,
      filterChoices = [],
      placeholder,
      value,
    } = this.props

    const { showFilters } = this.state

    const selectedDropdownIcon = this.dropdownIcons[dropdownIcon]

    const SearchButton = () => (
      <button
        type="button"
        onClick={() => onSubmit && onSubmit(value)}
        className={cx(['filterable-search__btn'])}
      >
        {buttonLabel}
      </button>
    )

    const DropdownToggle = () => (
      <button
        type="button"
        className={cx(['filterable-search__dropdown-toggle'])}
        onClick={() =>
          this.setState(prevState => ({
            showFilters: !prevState.showFilters,
          }))
        }
      >
        <div
          className={cx(['filterable-search__dropdown-toggle-icon'])}
          dangerouslySetInnerHTML={{ __html: selectedDropdownIcon }}
        />
      </button>
    )

    const DropdownMenu = ({ show }) => (
      <ul
        className={cx([
          'filterable-search__dropdown-menu',
          { 'filterable-search__dropdown-menu--open': show },
        ])}
      >
        {filterChoices.map(({ label, value, checked }, index) => (
          <li className={cx(['filterable-search__dropdown-menu-item'])} key={value}>
            {/* eslint-disable-next-line */}
            <label
              className={cx(['filterable-search__dropdown-menu-item-label'])}
              htmlFor={`${value}${index}`}
            >
              <Checkbox
                name={`${value}${index}`}
                checked={checked}
                onChange={() => onFilterChange({ label, value, checked })}
              />
              <span className={cx(['filterable-search__dropdown-menu-item-label-text'])}>
                {label}
              </span>
            </label>
          </li>
        ))}
      </ul>
    )

    return (
      <div className={cx(['filterable-search'])}>
        <TextField
          placeholder={placeholder}
          onChange={value => onInputChange(value)}
          onKeyDown={ev => {
            if (ev.keyCode === 13 && onSubmit) onSubmit(ev.target.value)
          }}
          value={value}
        />
        {onSubmit && <SearchButton />}
        <div ref={this.dropdownMenuRef}>
          <DropdownToggle />
          <DropdownMenu show={showFilters} />
        </div>
      </div>
    )
  }
}

FilterableSearch.propTypes = {
  buttonLabel: PropTypes.string,
  dropdownIcon: PropTypes.oneOf(['chevronDown', 'filter']),
  onFilterChange: PropTypes.func.isRequired,
  onInputChange: PropTypes.func,
  onSubmit: PropTypes.func,
  value: PropTypes.string.isRequired,
  filterChoices: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.string,
      checked: PropTypes.bool,
    }),
  ).isRequired,
  placeholder: PropTypes.string,
}

FilterableSearch.defaultProps = {
  buttonLabel: 'Search',
  dropdownIcon: 'chevronDown',
  placeholder: 'Search',
  onSubmit: undefined,
  onInputChange: undefined,
}

export default FilterableSearch
