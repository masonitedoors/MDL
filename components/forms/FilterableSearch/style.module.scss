@import "styles/index";

.filterable-search {
  position: relative;
  font-family: $font-family;
  display: flex;
  align-items: center;
  width: 100%;

  $height: 36px;

  &__btn {
    display: inline-block;
    height: $height;
    border: 0;
    font-family: $font-family;
    font-size: 12px;
    color: $color-white;
    padding: $size-md;
    background: $color-blue;
    cursor: pointer;
  }

  &__dropdown-toggle {
    display: inline-flex;

    @include square($height);

    border: 0;
    align-items: center;
    background: $color-blueberry;
    border-radius: 0 2px 2px 0;
    cursor: pointer;
  }

  &__dropdown-toggle-icon {
    @include square($size-lg);

    fill: white;
    margin: auto;
    pointer-events: none;
  }

  &__dropdown-menu {
    position: absolute;
    top: 100%;
    right: 0;
    margin: 0;
    padding: $size-lg;
    background: $color-white;
    box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.15);
    list-style: none;
    pointer-events: none;
    opacity: 0;
    z-index: 10;

    &--open {
      opacity: 1;
      pointer-events: all;
    }
  }

  &__dropdown-menu-item {
    margin: 0 0 $size-sm;
    padding: 0;

    &:last-of-type {
      margin-bottom: 0;
    }
  }

  &__dropdown-menu-item-label {
    display: flex;
    align-items: center;
  }

  &__dropdown-menu-item-label-text {
    margin-left: $size-md;
  }
}
