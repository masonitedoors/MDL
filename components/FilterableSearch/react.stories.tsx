import React from "react";
import { Story, Meta } from "@storybook/react";
import FilterableSearch from "./react";

const Template = (args) => <FilterableSearch {...args} />;

export default {
  title: "FilterableSearch",
  component: FilterableSearch,
} as Meta;

export const Primary = Template.bind({});

Primary.args = {
  children: "Lorem Button",
  filterChoices: [
    {
      label: "Filter 1",
      value: "FILTER1",
      checked: true,
    },
    {
      label: "Filter 2",
      value: "FILTER2",
      checked: false,
    },
  ],
  onFilterChange: () => {},
  placeholder: "",
  value: "",
};
