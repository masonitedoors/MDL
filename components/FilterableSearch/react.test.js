import React from 'react'
import { mount } from 'enzyme'
import { JSDOM } from 'jsdom'
import FilterableSearch from './react'

const jsdom = new JSDOM('<!doctype html><html><body><div id="app">Mount here</div></body></html>')
const { window } = jsdom

describe('FilterableSearch', () => {
  let props
  let wrapper
  let instance

  beforeEach(() => {
    props = {
      onFilterChange: jest.fn(),
      onInputChange: jest.fn(),
      onSubmit: jest.fn(),
      value: '',
      filterChoices: [
        { label: 'Filter 1', value: 'FILTER1', checked: false },
        { label: 'Filter 2', value: 'FILTER2', checked: false },
      ],
    }

    wrapper = mount(<FilterableSearch {...props} />, {
      attachTo: window.document.querySelector('#app'),
    })

    instance = wrapper.instance()
  })

  describe('with snapshots', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('with user actions', () => {
    it('Should have an initial state with the "showFilters" set to "false"', () => {
      expect(instance.state.showFilters).toBe(false)
    })

    it('Should set the showFilters in the state to be false when the dropdownToggle button is clicked two times', () => {
      const dropDownToggleButton = wrapper
        .find('DropdownToggle')
        .find('button')
        .at(0)

      dropDownToggleButton.simulate('click')
      expect(instance.state.showFilters).toBeTruthy()

      dropDownToggleButton.simulate('click')
      expect(instance.state.showFilters).toBe(false)
    })

    it('should trigger search action when the user hits enter in input or clicks SearchButton', () => {
      expect(props.onSubmit).toBeCalledTimes(0)
      wrapper.find('SearchButton').simulate('click')
      expect(props.onSubmit).toBeCalledTimes(1)
      wrapper.find('.text-field__input').simulate('keyDown', { keyCode: 1 }) // not Enter key
      wrapper.find('.text-field__input').simulate('keyDown', { keyCode: 13 }) // Enter key
      expect(props.onSubmit).toBeCalledTimes(2)
    })
  })
})
