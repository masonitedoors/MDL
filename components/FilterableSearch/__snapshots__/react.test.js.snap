// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterableSearch with snapshots should render correctly 1`] = `
<FilterableSearch
  buttonLabel="Search"
  dropdownIcon="chevronDown"
  filterChoices={
    Array [
      Object {
        "checked": false,
        "label": "Filter 1",
        "value": "FILTER1",
      },
      Object {
        "checked": false,
        "label": "Filter 2",
        "value": "FILTER2",
      },
    ]
  }
  onFilterChange={[MockFunction]}
  onInputChange={[MockFunction]}
  onSubmit={[MockFunction]}
  placeholder="Search"
  value=""
>
  <div
    className="filterable-search"
  >
    <TextField
      align="left"
      disabled={false}
      error={false}
      helper=""
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Search"
      readonly={false}
      removeMargin={false}
      type="text"
      value=""
      variant="dark"
    >
      <div
        className="text-field text-field--dark"
      >
        <div
          className="text-field__label"
        />
        <div
          className="text-field__helper"
        />
        <input
          className="text-field__input"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Search"
          readOnly={false}
          type="text"
          value=""
        />
        <TrailingIcon>
          <div
            className="text-field__icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": undefined,
              }
            }
          />
        </TrailingIcon>
      </div>
    </TextField>
    <SearchButton>
      <button
        className="filterable-search__btn"
        onClick={[Function]}
        type="button"
      >
        Search
      </button>
    </SearchButton>
    <div>
      <DropdownToggle>
        <button
          className="filterable-search__dropdown-toggle"
          onClick={[Function]}
          type="button"
        >
          <div
            className="filterable-search__dropdown-toggle-icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\"><title>chevron-down</title><path d=\\"M18.7 8.3c-0.4-0.4-1-0.4-1.4 0l-5.3 5.3-5.3-5.3c-0.4-0.4-1-0.4-1.4 0s-0.4 1 0 1.4l6 6c0.2 0.2 0.5 0.3 0.7 0.3s0.5-0.1 0.7-0.3l6-6c0.4-0.4 0.4-1 0-1.4z\\"></path></svg>",
              }
            }
          />
        </button>
      </DropdownToggle>
      <DropdownMenu
        showFilters={false}
      >
        <ul
          className="filterable-search__dropdown-menu"
        >
          <li
            className="filterable-search__dropdown-menu-item"
            key="FILTER1"
          >
            <label
              className="filterable-search__dropdown-menu-item-label"
              htmlFor="FILTER10"
            >
              <Checkbox
                checked={false}
                disabled={false}
                name="FILTER10"
                onChange={[Function]}
              >
                <div
                  aria-checked={false}
                  className="checkbox"
                  onClick={[Function]}
                  onKeyPress={[Function]}
                  role="checkbox"
                  tabIndex={0}
                >
                  <input
                    checked={false}
                    className="checkbox-input"
                    readOnly={true}
                    type="checkbox"
                  />
                  <div
                    className="checkbox-svg-wrapper"
                  >
                    <svg
                      className="checkbox-svg-icon"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\"><title>check</title><path d=\\"M20.7 5.3c-0.4-0.4-1-0.4-1.4 0l-10.3 10.3-4.3-4.3c-0.4-0.4-1-0.4-1.4 0s-0.4 1 0 1.4l5 5c0.2 0.2 0.4 0.3 0.7 0.3s0.5-0.1 0.7-0.3l11-11c0.4-0.4 0.4-1 0-1.4z\\"></path></svg>",
                        }
                      }
                    />
                  </div>
                </div>
              </Checkbox>
              <span
                className="filterable-search__dropdown-menu-item-label-text"
              >
                Filter 1
              </span>
            </label>
          </li>
          <li
            className="filterable-search__dropdown-menu-item"
            key="FILTER2"
          >
            <label
              className="filterable-search__dropdown-menu-item-label"
              htmlFor="FILTER21"
            >
              <Checkbox
                checked={false}
                disabled={false}
                name="FILTER21"
                onChange={[Function]}
              >
                <div
                  aria-checked={false}
                  className="checkbox"
                  onClick={[Function]}
                  onKeyPress={[Function]}
                  role="checkbox"
                  tabIndex={0}
                >
                  <input
                    checked={false}
                    className="checkbox-input"
                    readOnly={true}
                    type="checkbox"
                  />
                  <div
                    className="checkbox-svg-wrapper"
                  >
                    <svg
                      className="checkbox-svg-icon"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\"><title>check</title><path d=\\"M20.7 5.3c-0.4-0.4-1-0.4-1.4 0l-10.3 10.3-4.3-4.3c-0.4-0.4-1-0.4-1.4 0s-0.4 1 0 1.4l5 5c0.2 0.2 0.4 0.3 0.7 0.3s0.5-0.1 0.7-0.3l11-11c0.4-0.4 0.4-1 0-1.4z\\"></path></svg>",
                        }
                      }
                    />
                  </div>
                </div>
              </Checkbox>
              <span
                className="filterable-search__dropdown-menu-item-label-text"
              >
                Filter 2
              </span>
            </label>
          </li>
        </ul>
      </DropdownMenu>
    </div>
  </div>
</FilterableSearch>
`;
